### PR TITLE
SQLAlchemy instrumentor populates span after sqlcomment creation, not before

### PR DESCRIPTION
**Pull request by tammy-baylis-swi**
# Description

Re-orders database query span population by SQLAlchemy instrumentor so that `db.statement` span attribute includes any sqlcomments generated.

Fixes https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2938

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Added unit test. Locally installed this version of opentelemetry-instrumentation-sqlalchemy and upstream dependencies. Instrumented a Flask app that queries MySQL (via mysql-connector driver) with general logs enabled to check sqlcomment (enabled as per [doc](https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/sqlalchemy/sqlalchemy.html#sqlcommenter)). Spans were exported to a locally-running otel-collector to inspect attributes including `db.statement`.

After changes, example span from SQLAlchemy-with-mysql-connector query with updated db.statement:

```
InstrumentationScope opentelemetry.instrumentation.sqlalchemy 0.49b0.dev
Span #0
    Trace ID       : 3b654158a48190e458f6c4e15e125814
    Parent ID      : ed65cfad7167aecd
    ID             : 69d58000ef348179
    Name           : connect
    Kind           : Client
    Start time     : 2024-10-29 22:09:35.42945621 +0000 UTC
    End time       : 2024-10-29 22:09:35.466973419 +0000 UTC
    Status code    : Unset
    Status message :
Attributes:
     -> net.peer.name: Str(mysql-world-db)
     -> net.peer.port: Int(3306)
     -> db.name: Str(world)
     -> db.user: Str(world)
     -> db.system: Str(mysql)
Span #1
    Trace ID       : 3b654158a48190e458f6c4e15e125814
    Parent ID      : ed65cfad7167aecd
    ID             : b62c6e3e794e9f05
    Name           : SELECT world
    Kind           : Client
    Start time     : 2024-10-29 22:09:35.469197169 +0000 UTC
    End time       : 2024-10-29 22:09:35.469996085 +0000 UTC
    Status code    : Unset
    Status message :
Attributes:
     -> db.statement: Str(SELECT city.id, city.name, city.district, city.population, city.countrycode
FROM city
WHERE city.id = %(id_1)s /*db_driver='mysqlconnector',db_framework='sqlalchemy%%3A2.0.36',traceparent='00-3b654158a48190e458f6c4e15e125814-b62c6e3e794e9f05-01'*/)
     -> db.system: Str(mysql)
     -> net.peer.name: Str(mysql-world-db)
     -> net.peer.port: Int(3306)
     -> db.name: Str(world)
     -> db.user: Str(world)
```

which matches sqlcomment in corresponding MySQL general log:

```
2024-10-29T22:09:35.469520Z	  240 Query	SELECT city.id, city.name, city.district, city.population, city.countrycode
FROM city
WHERE city.id = '1818' /*db_driver='mysqlconnector',db_framework='sqlalchemy%%3A2.0.36',traceparent='00-3b654158a48190e458f6c4e15e125814-b62c6e3e794e9f05-01'*/
```

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [X] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [X] Followed the style guidelines of this project
- [X] Changelogs have been updated
- [X] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py [hunk 1]
```diff
@@ -218,11 +218,6 @@ def _before_cur_exec(
             kind=trace.SpanKind.CLIENT,
         )
         with trace.use_span(span, end_on_exit=False):
-            if span.is_recording():
-                span.set_attribute(SpanAttributes.DB_STATEMENT, statement)
-                span.set_attribute(SpanAttributes.DB_SYSTEM, self.vendor)
-                for key, value in attrs.items():
-                    span.set_attribute(key, value)
             if self.enable_commenter:
```
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py:221]: Does this code had any effect before the change? Like the changed statement is then reused later?

Comment (tammy-baylis-swi) [reply instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py:221]: I don't think it's reused. This method `_before_cur_exec` (that calls [_add_sql_comment](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/07c3324a3bcf80ee716f90b2456f59af08b685fa/opentelemetry-instrumentation/src/opentelemetry/instrumentation/sqlcommenter_utils.py#L19-L30) below on new L243) is only registered to SQLAlchemy [before_cursor_execute](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.ConnectionEvents.before_cursor_execute) via SQLAlchemy [native event registration](https://docs.sqlalchemy.org/en/20/core/event.html#event-registration). The executed statement that includes sqlcomment gets logged by db server (if configured). Afaik, the query result returned to client does not include the sqlcomment (or any comment) so it doesn't get passed back that way.

Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py:221]: If it's not reused maybe we can move its calculation under `span.is_recording`?

Comment (tammy-baylis-swi) [reply instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py]: Yes! Updated in [e3ddea9](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2937/commits/e3ddea996242d2466e56cd78a061b945f810b288). Still works to update `db.statement` and add sql comment.

**Review by lzchen — APPROVED** [review]

**Pull request merged**